shared/packages/worker/src/worker/accessorHandlers/s3.ts

```diff
@@ -0,0 +1,608 @@
+	}
+	checkHandleWrite(): AccessorHandlerCheckHandleWriteResult {
+		const defaultResult = defaultCheckHandleWrite(this.accessor)
+		if (defaultResult) return defaultResult
+		return { success: true }
+	}
+	async checkPackageReadAccess(): Promise<AccessorHandlerCheckPackageReadAccessResult> {
+		const s3 = await this.getS3Client()
+
+		const response = await s3.fileExists(this.fullPath)
+
+		if (response.exists) {
+			return { success: true }
+		} else {
+			return { success: false, reason: response.reason, knownReason: response.knownReason }
+		}
+	}
+	async tryPackageRead(): Promise<AccessorHandlerTryPackageReadResult> {
+		const s3 = await this.getS3Client()
+
+		const response = await s3.getFileInfo(this.fullPath)
+
+		if (response.success) {
+			return { success: true }
+		} else {
+			return {
+				success: false,
+				reason: response.reason,
+				knownReason: response.knownReason,
+				packageExists: response.packageExists,
+			}
+		}
+	}
+	async checkPackageContainerWriteAccess(): Promise<AccessorHandlerCheckPackageContainerWriteAccessResult> {
+		return { success: true }
+	}
+	private async checkPackageContainerReadAccess(): Promise<AccessorHandlerRunCronJobResult> {
+		return { success: true }
+	}
+	async getPackageActualVersion(): Promise<Expectation.Version.S3Resource> {
+		const s3 = await this.getS3Client()
+
+		const response = await s3.getFileInfo(this.fullPath)
+
+		if (response.success) {
+			return {
+				type: Expectation.Version.Type.S3_RESOURCE,
+				fileSize: response.fileInfo.size,
+				modifiedDate: response.fileInfo.modified,
+			}
+		} else throw new Error(`getPackageActualVersion: ${response.reason.user}: ${response.reason.tech}`)
+	}
+	async ensurePackageFulfilled(): Promise<void> {
+		await this.fileHandler.clearPackageRemoval(this.filePath)
+	}
+	async removePackage(reason: string): Promise<void> {
+		await this.fileHandler.handleRemovePackage(this.filePath, this.packageName, reason)
+	}
+	async getPackageReadStream(): Promise<PackageReadStream> {
+		const s3 = await this.getS3Client()
+
+		try {
+			const abortController = new AbortController()
+			const response = await s3.readFile(this.fullPath, abortController.signal)
+
+			return {
+				readStream: Readable.fromWeb(response),
+				cancel: () => {
+					try {
+						abortController.abort()
+					} catch (e: any) {
+						this.worker.logger.error(
+							`getPackageReadStream: Error when aborting S3 stream: ${
+								e && 'message' in e ? e.message : JSON.stringify(e)
+							}`
+						)
+					}
+				},
+			}
+		} catch (e: any) {
+			this.worker.logger.error(
+				`getPackageReadStream: Error when downloading S3 stream: ${
+					e && 'message' in e ? e.message : JSON.stringify(e)
+				}`
+			)
+
+			throw e
+		}
+	}
+
+	async putPackageStream(sourceStream: NodeJS.ReadableStream): Promise<PutPackageHandler> {
+		const s3 = await this.getS3Client()
+
+		const fullPath = this.workOptions.useTemporaryFilePath ? this.temporaryFilePath : this.fullPath
```

Comment:
I just realized that `useTemporaryFilePath` might not be needed for S3 storage.
It is used when writing to files on disk to ensure atomic operations, but [since S3 operations are atomic by definition](https://stackoverflow.com/questions/23764683/are-writes-to-amazon-s3-atomic-all-or-nothing), I think we can omit any uses of `useTemporaryFilePath` completely.